apps/web/utils/fetcher.ts

```diff
@@ -1,9 +1,62 @@
-export default async function fetcher<T>(input: RequestInfo, init?: RequestInit): Promise<T> {
-	const response = await fetch(input, init);
+// export default async function fetcher<T>(input: RequestInfo, init?: RequestInit): Promise<T> {
+// 	const response = await fetch(input, init);
+
+// 	if (!response.ok) {
+// 		throw new Error(`An error occurred while fetching the data: ${response.statusText}, ${await response.text()}`);
+// 	}
+
+// 	return response.json() as Promise<T>;
+// }
+
+export interface FetcherResult<T> {
+	data?: T | null;
+	error?: any; // You can specify a more specific error type if needed
+}
+
+export const convertResponse = async (response: Response) => {
+	let responseBody;
+
+	const contentType = response.headers.get("content-type")?.toLowerCase() ?? "";
+
+	if (contentType.includes("application/json")) {
+		responseBody = await response.json();
+	} else {
+		responseBody = await response.text();
+	}
 
 	if (!response.ok) {
-		throw new Error(`An error occurred while fetching the data: ${response.statusText}, ${await response.text()}`);
+		// If the response is not OK, return the error
+		// returns response with { data: null, error: {} }
+		if (responseBody.data === null && responseBody.error !== null) {
+			return responseBody as Response;
```

Comment:
https://github.com/stamford-syntax-club/course-compose/blob/main/apps/web/types/errors.ts
```ts
return responseBody as ErrorResponse
```
(This is just the format that backend uses to return error, no worries XD)

Reply:
Noted